aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1056,6 +1056,150 @@ std::tuple<Tensor,optional<int64_t>> masked_fill_scalar_batch_rule(
   return std::make_tuple(result, 0);
 }
 
+std::tuple<Tensor,optional<int64_t>> index_fill_int_scalar_batch_rule_impl(
+    Tensor & self, optional<int64_t> self_bdim,
+    int64_t dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Scalar & value,
+    const bool inplace) {
+
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor index_ = moveBatchDimToFront(index, index_bdim);
+    dim = maybe_wrap_dim(dim, self_.dim());
+
+    if (inplace && !self_bdim.has_value()) {
+      vmapIncompatibleInplaceError("index_fill_");
+    }
+
+    if (!index_bdim) {
+      self_.unsqueeze_(-1).index_fill_(dim + 1, index_, value).squeeze_(-1);
+      return std::make_tuple(self_, 0);
+    }
+
+    auto batch_size = get_bdim_size2(self, self_bdim, index, index_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+
+    if (inplace) {
+      // Do for-loop for in-place because we cannot reshape
+      // `self_` having an incompatible stride without copying
+      for (const auto i : c10::irange(0, batch_size)) {
+        const auto& self_slice = self_.select(0, i);
+        const auto& index_slice = index_.select(0, i);
+        self_slice.index_fill_(
+          dim,
+          index_slice,
+          value
+        );
+      }
```

Comment:
nit: Let's do an early return here so it's clear that we're done with this case.

```
if (inplace) {
  ...
  return std::make_tuple(self_, 0);
}
// !inplace case
```